## try语句块和异常处理

- 在真实的程序中，应该把内部执行过程和用户交互的代码分开。

-  ```cpp
  try{
    program_statements
  }catch(exception_declaration){
    handler_statements
  }catch(exception_declaration){
    handler_statements
  }
  ```

- try语句块的program_statements组成程序的正常逻辑。 可以有包括声明在内的任意C++语句。

  - 一如往常，try语句块内声明的变量在块外无法访问。catch语句也不能访问。
  - program_statements内应该包含了throws语句，遇到异常时要抛出异常。

- catch语句块的exception_declaration是异常声明。要和throws抛出的异常名匹配。

  - catch完成后，执行最后一个catch之后的那条语句。
  - 如果没有找到匹配的catch语句，就会跳转到名为terminate的标准库函数。 一般情况，执行该函数将导致程序非正常退出。

## 标准异常

- 分布在四个头文件中

  - exception定义了最通用的异常类exception。只报告异常的发生，不提供额外信息。

  - stdexcept顶一个几种常用异常类

  - | 异常名              | 意义                      |
    | ---------------- | ----------------------- |
    | exception        | 最常见问题                   |
    | runtime_error    | 运行时才会发现的问题              |
    | range_error      | 运行时错误：生成的结果超出有意义的值域范围   |
    | overflow_error   | 运行时的错误：计算上溢             |
    | underflow_error  | 运行时错误：计算下溢              |
    | logic_error      | 程序逻辑错误...               |
    | domain_error     | 逻辑错误：参数对应的结果值不存在        |
    | invalid_argument | 逻辑错误：无效参数               |
    | length_error     | 逻辑错误：试图创建一个超出该类型最大长度的对象 |
    | out_of_range     | 逻辑错误：使用一个超出有效范围的值       |

    new头文件定义了bad_alloc异常类型。

  - type_info头文件定义了bad_cast异常类型。

- 对于exception , bad_alloc ,bad_cast对象，只能以默认初始化的方式初始化（也就是...不需要初始化）.， 不允许为这些对象提供初始值。

- 而对于其他异常类型，应当使用string或者C风格字符串初始化这些类型的对象，并且**不允许使用默认初始化的方式**。

  (初始值应含有与错误相关的信息)

- 异常类型只有一个名为what的成员函数， 该函数没有任何参数， 返回值是一个const char* 。 目的是提供关于异常的文本信息。

  - 如果异常类型有字符串初始值，则返回该字符串，否则由编译器决定返回的值。